cranelift/fuzzgen/src/function_generator.rs

```diff
@@ -542,6 +542,161 @@ const OPCODE_SIGNATURES: &'static [(
+    builder: &mut FunctionBuilder,
+    source_block: Block,
+) -> Result<()> {
+    let (block, args) = fgen.generate_target_block(builder, source_block)?;
+    builder.ins().jump(block, &args[..]);
+    Ok(())
+}
+
+/// Generates a br_table into a random block
+fn insert_br_table(
+    fgen: &mut FunctionGenerator,
+    builder: &mut FunctionBuilder,
+    source_block: Block,
+) -> Result<()> {
+    let var = fgen.get_variable_of_type(I32)?; // br_table only supports I32
+    let val = builder.use_var(var);
+
+    let target_blocks = fgen.resources.forward_blocks_without_params(source_block);
+    let default_block = *fgen.u.choose(target_blocks)?;
+
+    // We can still select a backwards branching jump table here!
+    let tables = fgen.resources.forward_jump_tables(builder, source_block);
+    let jt = *fgen.u.choose(&tables[..])?;
+    builder.ins().br_table(val, default_block, jt);
+    Ok(())
+}
+
+/// Generates a brz/brnz into a random block
+fn insert_br(
+    fgen: &mut FunctionGenerator,
+    builder: &mut FunctionBuilder,
+    source_block: Block,
+) -> Result<()> {
+    let (block, args) = fgen.generate_target_block(builder, source_block)?;
+
+    let condbr_types = [I8, I16, I32, I64, I128, B1];
+    let _type = *fgen.u.choose(&condbr_types[..])?;
+    let var = fgen.get_variable_of_type(_type)?;
+    let val = builder.use_var(var);
+
+    if bool::arbitrary(fgen.u)? {
+        builder.ins().brz(val, block, &args[..]);
+    } else {
+        builder.ins().brnz(val, block, &args[..]);
+    }
+
+    // After brz/brnz we must generate a jump
+    insert_jump(fgen, builder, source_block)?;
+    Ok(())
+}
+
+fn insert_bricmp(
+    fgen: &mut FunctionGenerator,
+    builder: &mut FunctionBuilder,
+    source_block: Block,
+) -> Result<()> {
+    let (block, args) = fgen.generate_target_block(builder, source_block)?;
+    let cond = *fgen.u.choose(IntCC::all())?;
+
+    let bricmp_types = [
+        I8, I16, I32,
+        I64,
+        // I128 - TODO: https://github.com/bytecodealliance/wasmtime/issues/4406
+    ];
+    let _type = *fgen.u.choose(&bricmp_types[..])?;
+
+    let lhs_var = fgen.get_variable_of_type(_type)?;
+    let lhs_val = builder.use_var(lhs_var);
+
+    let rhs_var = fgen.get_variable_of_type(_type)?;
+    let rhs_val = builder.use_var(rhs_var);
+
+    builder
+        .ins()
+        .br_icmp(cond, lhs_val, rhs_val, block, &args[..]);
+
+    // After bricmp's we must generate a jump
+    insert_jump(fgen, builder, source_block)?;
+    Ok(())
+}
+
+fn insert_switch(
+    fgen: &mut FunctionGenerator,
+    builder: &mut FunctionBuilder,
+    source_block: Block,
+) -> Result<()> {
+    let _type = *fgen.u.choose(&[I8, I16, I32, I64, I128][..])?;
+    let switch_var = fgen.get_variable_of_type(_type)?;
+    let switch_val = builder.use_var(switch_var);
+
+    let default_block = {
+        let target_blocks = fgen.resources.forward_blocks_without_params(source_block);
+        *fgen.u.choose(target_blocks)?
```

Comment:
Seems unfortunate that `insert_switch` can't generate backward branches, but I guess that's fine.

Reply:
Yeah, we should probably include that

Reply:
I don't mind leaving this as something to come back to later though. If you want to just add a TODO comment to `insert_switch` that it would be nice to also check backward branches, that's plenty.